### ISSUE TITLE
Add support for JUnit5

### ISSUE THREAD

**Issue by dadoonet** [randomizedtesting/randomizedtesting#256]
Would be nice to add support for JUnit5.

**Comment by dweiss** [comment]
Would be. Feel free to provide a patch. :)

**Comment by dadoonet** [comment]
I started to see if I can do it (instead of opening an issue) but I've been blocked almost immediately with the ant part of the project. Some Java9 issues I guess... I might look at it indeed. :)

**Comment by dweiss** [comment]
To be honest I haven't had the need to use JUnit5, so my knowledge of it is very limited. I know they've changed the internals a lot, so it may be time-consuming refactoring. If there is not clear benefit I don't think I'll invest my time in doing this... there's just not enough of it anymore.

**Comment by dadoonet** [comment]
I can totally understand. That's what makes open source so funny: DIY! :p 
I'll update the ticket if I make some progress. (Well, I do: building with Java8 works well)

**Comment by dadoonet** [comment]
@dweiss Correct me if I'm wrong but the maven plugin "just" delegates to the ant one.
If so I believe the first step is to create an `junit5-ant` module. But for that we need to wait for: https://github.com/apache/ant/pull/60 to be merged and released.

**Comment by jprante** [comment]
I'm also interested in an update. JUnit 5 is a big step.

Randomized testing in JUnit5 might look like this https://github.com/qala-io/datagen/tree/master/junit5

Because I'm not interested in Maven or Ant, a pure Gradle solution is a viable alternative to me. Gradle 4.6 will support JUnit5 https://github.com/gradle/gradle/issues/828#issuecomment-362594770 

There is a first step https://github.com/gradle/gradle/pull/4116

Ordering sequence of tests might be important https://github.com/junit-team/junit5/issues/13#issuecomment-323526272 was considered for JUnit 5.1 which was released yesterday  https://github.com/junit-team/junit5/releases/tag/r5.1.0 but issue is still open.

Parallelized executions is another issue which is not settled in JUnit 5 https://github.com/junit-team/junit5/issues/60 but would be very interesting

**Comment by atorok** [comment]
I'm also approaching this from the Gradle perspective. 
At some point Gradle will make it impossible to replace the testing task as the plugin currently does. 
Also by replacing the testing task, one misses out on features around build scan and reporting. 
As part of elastic/elasticsearch#31496 I was looking into ways to be able to run the randomzied elasticsearch tests with Gradle. The major feature loss is around being able to randomized test case order, that for junit 4 is handled by Gradle, so Gradle could deal with the pseudo randomizing the test execution order - gradle/gradle#5760. Gradle does support junit 5 now as @oehme points it out, and we could use `junit-vintage-engine` to run the tests unchanged with Gradle and benefit from both Gradle and junit5 without code changes. 
To also get the randomized test order and other features, we would need to implement a test engine based on the junit-vintage-engine that has all the randomization features ? What do you think about that @dweiss ?

**Comment by dweiss** [comment]
I honestly don't know, haven't touched JUnit5 yet.

**Comment by Nick-Minutello** [comment]
I am using JUnit 5 and am sorely missing randomized testing.
It doesn't look like a small task to support Junit 5.
Is there any appetite to do it?
What are people using instead?

**Comment by dweiss** [comment]
You're correct - I don't think it's a small task, especially if you wanted to fit into JUnit5's new architecture more nicely... For now, I'm happy working with JUnit4 so there's no real itch to upgrade (and rewrite thousands of tests)...

**Comment by Nick-Minutello** [comment]
Do you even think it is possible without starting from scratch ?
I am wondering if it makes sense to start with a very cut-down implementation for junit5 & port features bit by bit?

**Comment by dweiss** [comment]
It depends on the direction, really. Most people think of JUnit5 in terms of Jupiter - the default TestEngine implementation. You could (easily?) port randomizedtesting's runner as a separate TestEngine but then none of jupiter's extensions would work, which is not much different than running JUnit5 tests with the vintage TestEngine... 

A more ambitious plan would be to try to hack into jupiter to simulate what the custom runner currently does (primarily deterministic randomized context/seed providers for tests, hooks, etc., then thread leak detection and listeners, perhaps). 

Last time I did some initial experiments [1], I couldn't make it work but I'm sure a lot has changed since then.

I think it'd be a fun project if you decide to spend some time on it. It's self-contained, which makes it even nicer. You can copy/paste or reuse whatever you need but you also don't have to - it's not about the implementation, it's about the functionality it has to offer. I'd start with the order in which the examples [2] are laid out - I tried to make it progress from easier/ more fundamental features to more esoteric ones.

[1] https://github.com/randomizedtesting/randomizedtesting/commits/junit5/
[2] https://github.com/randomizedtesting/randomizedtesting/tree/master/examples/maven/src/main/java/com/carrotsearch/examples/randomizedrunner

**Comment by dweiss** [comment]
I'll close this issue. The jupiter version works quite well for me, actually - I've ported it to a few projects and - while not automatic - it does work fairly well, even when combined with other jupiter facilities [1]. Thank you for the nudge.

[1] https://github.com/randomizedtesting/randomizedtesting-jupiter.

**Comment by dadoonet** [comment]
Thank you so much Dawid!
